### DISCUSSION TITLE
How can I make go pick the right version when using 'dev'?

### DISCUSSION THREAD

**Discussion by fredrikaverpil** [pkgxdev/pkgx#876] [Help] [answered]
I've got a project with a `go.mod` file which specifies `go 1.17`. Should this not be detected automatically (when `dev` has been run in the project) and make sure Go version 1.17 is active?

What's a good way to tell this specific project of mine to use Go 1.17?

**Comment by jhheider** [comment, marked as answer]
```yaml
dependencies:
  go.dev: ~1.17 # or, e.g., =1.17.2
```
looking at https://github.com/pkgxdev/pkgx/blob/7a488fd184b31ef4f96f64f152426b8fefe7ee6c/src/utils/devenv.ts#L57-L61, pkgx only adds `go.dev: '*'`, but doesn't parse go.mod. personally, i'm torn as to whether that is strictly a good move. while `go.mod` does define a go version, it's often possible it work with other go versions, and i can definitely see developers wanting to be able to switch/test more easily. usually, when in doubt, we err on the side of user-configurability.

what do you think? is the ability to set complex version requirements using `pkgx.yaml` worthwhile here?

**Reply by fredrikaverpil** [reply]
Good question. Let me ask some colleagues, gather some more input, and I'll get back to you with what we think.

**Reply by jhheider** [reply]
sounds like a plan. we've been trying to walk a fine balance here. as an example: we has a user who has to use a patched version of `yarn` for their project, but `pkgx` detects `yarn.lock` and adds our version to their environment. so, each invocation, they have to do `env -yarnpkg.com` to get correct functionality. i worry we'd be buying something like that with parsing `go.mod`.

but, of course, we do parse the `engines` key of `package.json`, though it often specifies semver ranges. just the various considerations at play.

**Reply by fredrikaverpil** [reply]
I think let's not do anything about this. I'm still new to both Go and pkgx and I can easily work around this by adding a [pkgx.yml file](https://docs.pkgx.sh/using-dev/dev#pkgx.yaml) if I actually want Go to pick another version.

I appreciate the conservative approach here.

**Reply by jhheider** [reply]
excellent. yes, it's a fine line between helping people work more easily and locking them into things they don't want.

**Reply by mycrEEpy** [reply]
Just for information, since Go 1.21 we have [toolchain management](https://go.dev/blog/toolchain) which basically makes it obsolete to have the "correct" Go version installed locally.

**Comment by Aditya-doid** [comment]
Good question..i iwait
